### ISSUE TITLE
Problems fetching bounce emails via $message->getAttachments();

### ISSUE THREAD

**Issue by protos1575** [kartolo/direct_mail#72]
I try to use AnalyzeBounceMail scheduler task, with your implementation. 
You obviously assume that the returning emails are attachments on the bounce emails returning from the server. In my case it is not. The returning emails are embedded in the email body. So I changed the  code for the processBounceMail to:
 
    /**
     * Process the bounce mail
     * @param Message $message the message object
     * @return bool true if bounce mail can be parsed, else false
     */
    private function processBounceMail($message)
    {
        /** @var Readmail $readMail */
        $readMail = GeneralUtility::makeInstance(Readmail::class);
        $midArray = array();
	$bouncedMail = $message->getMessageBody();
        // Find mail id
        $midArray = $readMail->find_XTypo3MID($bouncedMail);
        // Extract text content
        $cp = $readMail->analyseReturnError($message->getMessageBody());

        $res = $this->getDatabaseConnection()->exec_SELECTquery(
                'uid,email',
                'sys_dmail_maillog',
                'rid=' . intval($midArray['rid']) . ' AND rtbl="' .
                $this->getDatabaseConnection()->quoteStr($midArray['rtbl'], 'sys_dmail_maillog') . '"' .
                ' AND mid=' . intval($midArray['mid']) . ' AND response_type=0'
         );

         // only write to log table, if we found a corresponding recipient record
         if ($this->getDatabaseConnection()->sql_num_rows($res)) {
                $row = $this->getDatabaseConnection()->sql_fetch_assoc($res);
                $midArray['email'] = $row['email'];
                $insertFields = array(
                    'tstamp' => $GLOBALS['EXEC_TIME'],
                    'response_type' => -127,
                    'mid' => intval($midArray['mid']),
                    'rid' => intval($midArray['rid']),
                    'email' => $midArray['email'],
                    'rtbl' => $midArray['rtbl'],
                    'return_content' => serialize($cp),
                    'return_code' => intval($cp['reason'])
                );
                return $this->getDatabaseConnection()->exec_INSERTquery('sys_dmail_maillog', $insertFields);
            } else {
                return false;
            }
    }

This does the job for me. 
Can you explain your adoption, that the returning emails are **attachments** on the bounce emails?

**Comment by kartolo** [comment]
on my setup (postfix, dovecot) the bounce mail is split in 3 section:

1. Notification message

> I'm sorry to have to inform you that your message could not
be delivered to one or more recipients. It's attached below.
> 
> For further assistance, please send mail to postmaster.
> 
> If you do so, please include this problem report. You can
delete your own text from the attached returned message.
> 
>                    The mail system
> 
> <notexist@domain.tld>: user unknown

2. delivery report
> Reporting-MTA: dns; kartolo.local
> X-Postfix-Queue-ID: D9C1946CC4
> X-Postfix-Sender: rfc822; typo3@kartolo.dev
> Arrival-Date: Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
> 
> Final-Recipient: rfc822; notexist@kartolo.dev
> Original-Recipient: rfc822;notexist@kartolo.dev
> Action: failed
> Status: 5.1.1
> Diagnostic-Code: x-unix; user unknown

3. original email

can you attach some source code of the bounce mail?

source code of my bounce mail (postfix, dovecot on Ubuntu 16):
    Return-Path: <MAILER-DAEMON>
    Delivered-To: typo3@kartolo.dev
    Received: by kartolo.local (Postfix)
    	id E375D46CC6; Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
    Date: Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
    From: MAILER-DAEMON@kartolo.local (Mail Delivery System)
    Subject: Undelivered Mail Returned to Sender
    To: typo3@kartolo.dev
    Auto-Submitted: auto-replied
    MIME-Version: 1.0
    Content-Type: multipart/report; report-type=delivery-status;
    	boundary="D9C1946CC4.1473967668/kartolo.local"
    Content-Transfer-Encoding: 7bit
    Message-Id: <20160915192748.E375D46CC6@kartolo.local>
    
    This is a MIME-encapsulated message.
    
    --D9C1946CC4.1473967668/kartolo.local
    Content-Description: Notification
    Content-Type: text/plain; charset=us-ascii
    
    This is the mail system at host kartolo.local.
    
    I'm sorry to have to inform you that your message could not
    be delivered to one or more recipients. It's attached below.
    
    For further assistance, please send mail to postmaster.
    
    If you do so, please include this problem report. You can
    delete your own text from the attached returned message.
    
                       The mail system
    
    <notexist@kartolo.dev>: user unknown
    
    --D9C1946CC4.1473967668/kartolo.local
    Content-Description: Delivery report
    Content-Type: message/delivery-status
    
    Reporting-MTA: dns; kartolo.local
    X-Postfix-Queue-ID: D9C1946CC4
    X-Postfix-Sender: rfc822; typo3@kartolo.dev
    Arrival-Date: Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
    
    Final-Recipient: rfc822; notexist@kartolo.dev
    Original-Recipient: rfc822;notexist@kartolo.dev
    Action: failed
    Status: 5.1.1
    Diagnostic-Code: x-unix; user unknown
    
    --D9C1946CC4.1473967668/kartolo.local
    Content-Description: Undelivered Message
    Content-Type: message/rfc822
    Content-Transfer-Encoding: 7bit
    
    Return-Path: <typo3@kartolo.dev>
    Received: from kartolo.local (localhost [127.0.0.1])
    	by kartolo.local (Postfix) with ESMTP id D9C1946CC4
    	for <notexist@kartolo.dev>; Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
    X-Virus-Scanned: Debian amavisd-new at kartolo.local
    Received: from kartolo.local ([127.0.0.1])
    	by kartolo.local (kartolo.local [127.0.0.1]) (amavisd-new, port 10024)
    	with ESMTP id RSmKfHQBAest for <notexist@kartolo.dev>;
    	Thu, 15 Sep 2016 21:27:48 +0200 (CEST)
    Received: by kartolo.local (Postfix, from userid 33)
    	id EED6246CC3; Thu, 15 Sep 2016 21:27:47 +0200 (CEST)
    To: Test bounce <notexist@kartolo.dev>
    Subject: test Test Newsletter
    X-PHP-Originating-Script: 1000:SimpleMailInvoker.php
    Message-ID: <afbf91a0e60bc5526160f48323d5cba7@kartolo.dev>
    Date: Thu, 15 Sep 2016 19:27:47 +0000
    From: Ivan <ivan.kartolo@gmail.com>
    Reply-To: test <ivan.kartolo@gmail.com>
    MIME-Version: 1.0
    Content-Type: multipart/alternative;
     boundary="_=_swift_v4_1473967667_e37b471b7c2c1fa7c1d5cb0a75e64bbd_=_"
    X-Priority: 3 (Normal)
    X-TYPO3MID: MID2_t2-085f9803e9d00f263cf4c0704dd8389d
    X-Mailer: TYPO3
    
    
    --_=_swift_v4_1473967667_e37b471b7c2c1fa7c1d5cb0a75e64bbd_=_
    Content-Type: text/plain; charset=utf-8
    Content-Transfer-Encoding: quoted-printable
    
    Hello.
    
    This is default template for plain-text mail content
    
    
    =
    
    [Unrendered Content Element: textmedia ]
    
    
    
    
    
    --------------=
    --- FOOTER: ----------------------
    Unsubscribe link:
    http://example.com=
    .test/?id=3D3&cmd=3Dedit&aC=3Df70fd615&rU=3D2
    
    
    
    
    
    
    
    --_=_swift_v4_1473967667_e37b471b7c2c1fa7c1d5cb0a75e64bbd_=_
    Content-Type: text/html; charset=utf-8
    Content-Transfer-Encoding: quoted-printable
    
    <!DOCTYPE html>
    <html lang=3D"en">
    <head>
    
    <meta charset=3D"utf-8">=
    <title>TYPO3 CMS - Introduction Package: Test Newsletter</title>
    <meta na=
    me=3D"generator" content=3D"TYPO3 CMS">
    
    
    
    
    
    
    
    
    
    
    </h=
    ead>
    <body>
    
    <div id=3D"c165">
    =09
    =09=09
    
    =09
    =09=09
    =09=
    =09=09=09
    =09=09=09
    
    =09=09
    
    =09
    =09=09<h1>test content</h1>
    =
    =09
    =09
    =09
    =09
    =09
    =09
    =09
    =09
    
    
    
    =09=09
    
    =09=09=
    
    
    
    
    
    =09=09
    =09=09=09=09
    =09=09=09
    =09
    
    
    =09
    
    =09=
    <div class=3D"ce-textpic ce-right ce-intext">
    =09=09
    =09=09=09
    
    
    =
    
    
    
    
    
    
    
    
    
    =09=09
    
    =09=09<div class=3D"ce-bodytext">
    =
    =09=09=09
    =09=09=09<p>Nulla facilisi. Duis aliquet egestas purus in bland=
    it. Curabitur vulputate, ligula lacinia scelerisque tempor, lacus lacus orn=
    are ante, ac egestas est urna sit amet arcu. Class aptent taciti sociosqu a=
    d litora torquent per conubia nostra, per inceptos himenaeos. Sed molestie =
    augue sit amet leo consequat posuere. Vestibulum ante ipsum primis in fauci=
    bus orci luctus et ultrices posuere cubilia Curae; Proin vel ante a orci te=
    mpus eleifend ut et magna. Lorem ipsum dolor sit amet, consectetur adipisci=
    ng elit. Vivamus luctus urna sed urna ultricies ac tempor dui sagittis. In.=
    
    </p>
    <p>&nbsp;</p>
    =09=09</div>
    
    =09=09
    =09</div>
    
    =09
    =
    
    
    </div>
    
    
    
    
    
    
    </body>
    </html>
    
    --_=_swift_v4_1473967667_e37b471b7c2c1fa7c1d5cb0a75e64bbd_=_--
    
    
    --D9C1946CC4.1473967668/kartolo.local--

**Comment by protos1575** [comment]
In my case it looks like something like that:

`Von: WEB.DE Mailer Daemon
Gesendet: Montag, 11. September 2017 06:56:04 (UTC+01:00) Amsterdam, Berlin=
, Bern, Rom, Stockholm, Wien
An: prvs=3D0427339fab=3Dmail-bounces@matthies.de
Betreff: Mail delivery failed: returning message to sender

This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its recipi=
ents. This is a permanent error. The following address(es)
failed:

gernot.erichsen@online.de:
SMTP error from remote server for RCPT TO command, host: mx01.emig.kundense=
rver.de (217.72.192.66) reason: 550 Requested action not taken: mailbox una=
vailable



--- The header of the original message is following. ---

Received: from [212.227.15.17] ([212.227.15.17]) by mx-ha.web.de (mxweb010
 [212.227.15.17]) with ESMTPS (Nemesis) id 1N8HMi-1dMh9Q1TZA-014ldm for  <g=
ernot.erichsen@online.de>; Mon, 11 Sep 2017 06:56:04 +0200
Received: from fw01.matthies.de ([81.18.96.56]) by mx-ha.web.de (mxweb010
 [212.227.15.17]) with ESMTPS (Nemesis) id 1Mkof5-1dAGpn1TTz-00lj3R for  <g=
erichsen@web.de>; Mon, 11 Sep 2017 06:56:04 +0200
Received: from [81.18.99.20] (port=3D47640 helo=3Djjmmgt.matthies.de)
        by fw01.matthies.de with esmtp (Exim 4.82_1-5b7a7c0-XX)
        (envelope-from <mail-bounces@matthies.de>)
        id 1drGfk-0000c1-39
        for gerichsen@web.de; Mon, 11 Sep 2017 06:50:12 +0200
Received: by jjmmgt.matthies.de (Postfix, from userid 0)
        id EF0F45FBA1; Mon, 11 Sep 2017 06:50:12 +0200 (CEST)
To: "MICHAEL ERICHSEN              " <gerichsen@web.de>
Subject: JA! Sonderangebot. Akku-Bohrschrauber 18 V. Extrem handlich.
X-PHP-Originating-Script: 1002:SimpleMailInvoker.php
Message-ID: <9df4b7991e5449ae0f25279f9fff47a4@swift.generated>
Date: Mon, 11 Sep 2017 06:50:12 +0200
From: MATTHIES - einfach mehr drin <newsletter@matthies.de>
Reply-To: MATTHIES - einfach mehr drin <newsletter@matthies.de>
MIME-Version: 1.0
Content-Type: multipart/alternative;
 boundary=3D"_=3D_swift_v4_1505105412_39ce0cef66143607623303c59816b224_=3D_=
"
X-TYPO3MID: MID7054_f35263-fd6bbe0bd2cefe34627b726102b0dde5
X-Priority: 3 (Normal)
X-Mailer: TYPO3
Organization: MATTHIES - einfach mehr drin
X-UI-Out-Filterresults: notjunk:1;V01:K0:kl2wUSCGpYk=3D:fbwAyy1fzOnvdmzJloJ=
cXK
 H1tsJoL8Dary12nMQYQeRi5rj+UvwpDeL4Dx0cRGTu/8oyi54xZf8UIXspOkPLPgUHtfJYxHD
 D/knzD5a8SzkMIbxIN+4Kvq/4m6HRxG33Dl/ThpM8yaLZI7VBxzWvuDxMNxV0wVnXCHeBiXyB
 wZsiYPggBz/BiqJ+YRczIpZ6KgDwo4AJYhV13jHUYOREETzhHJVBUdnIz69Nx+yicoCfT9xhi
 ++YrMvyy8/lByRW22Z8Fv4BuW70cU4yeADr2CvjZ2679QsWIo3KRo/LCHTJsTvWwVPjYja0Yb
 fDdZsRS8jkrSL3/4Jebeg08Y1CTvWaFz8xMRISno894GzKlPL9Xk3Xkg5dxSQ+ZnhH7AFVC6t
 yFj2miuSuVXJ50aYuZL1bVTON6VP0B9NrTVHbdBBiX2EKB7MJGwWMO4PFFGH3ivs8bcW/I4+t
 xLkO09dAodJ+uWulJVLGtccQx/ok1Rbdl1hw/AyGGMZ7veY9acvluguK8aqbrT63gnTIsJD1a
 M8CiMr5xUCsgkrZ7cqgZeHDR43zxJxrNUH5PUxpF56SpLycpGInLshGo8FCU3qDwv+jvYhNNb
 Bi1Ssh4ohLCnGolCm23MbodqwAWTl7/nWKXihO6+RAm1XA+iMUOoDE4tfNnkWXqUwTrKhBWxm
 DccDyJt7CGqVdOFLzCRMcH2sdDFSbBdEB1+pZWdf6dZNK4y9KhIikwEIhlWW95nSC3ApFmsuM
 VDIj/bJwd2C9qVIqBKkaPKgx+PBFMwPIHY7SPJQ8RJJk+Z7uSAOYt/P2mTRR/k+GamnTBpRsH
 UaZ7zJ1eVCW/XulOrfroCIAzZ4LTJ5dQwxYQ8AHquIehmmnc1N+668EuYQ4dXCtfj0Acw/6bi
 0WNtjqdMEh5ZCT7Wj7Zj5U60C4Go4z/UaH/dr7kfUOiRbdsqWvd08gXoT1GZI8/s3334CbTwN
 +uAKW0cBSwFJXxnLw4LW1L/GY8y16ktD1pDIqI6KjW8/KVaOd0b8ukhhNCCIlhin7j9vacIEU
 lsIqrZ4+TYPiSosped+SWhuf0xXTF50DqQgHxovY+aphIDKoJfVUg/5+ayaLBV6WZkiE6Srii
 ekPTJTc/CD0sGqlmm6fwlNZIp7cAmGsqRqalNtZk7rgEJ+Rl1BrLr9MKOcxAdePsqRDcrSGm0
 cV0k9b8b4l0wCSwKUQHmQa9+aNxtrux9RaC+o3t4oGx4Ot/zbgPFk5rnIdjIyxO5qe/VOOl6/
 fJWahHeiD4mtTEDbHHTpJW8qGIMTFZS9SJJHU4vwou07QBvoPQPQ25dXISR8uwyfDJmwTCqTA
 oYSWucnGWIBYr9hz8jME18f+DvcbFjtcpGBY1YvtOBJe6pFcP71nhuVlMjSLCK6DEtyRl9kTW
 4oUJLG+ZYeqOWX85GviCw=3D=3D


`

**Comment by Kalabybis** [comment]
your solution produces a lot of notices if the message has no attachments. Check AnalyzeBounceMail->processBounceMail the first if statement:
`if (is_array($attachmentArray)) {`
Since Message->getAttachments has no $filename parameter set it will return all attachments which is always an array. Message.php:
`protected $attachments = array();`
$midArray stays empty which will generate notices while executing a SELECT query:
`
$res = $this->getDatabaseConnection()->exec_SELECTquery(
                'uid,email',
                'sys_dmail_maillog',
                'rid=' . intval($midArray['rid']) . ' AND rtbl="' .
                $this->getDatabaseConnection()->quoteStr($midArray['rtbl'], 'sys_dmail_maillog') . '"' .
                ' AND mid=' . intval($midArray['mid']) . ' AND response_type=0'
            );
`

**Comment by SventB** [comment]
@protos1575 I have this problem, too, the returned mails have no attachment. My TYPO3/mail system is running on a Hetzner server.

@Kalabybis I don't understand your answer, it seems to be off topic. It doesn't help @protos1575 and @kartolo to find a solution for the problem, that some bounce mails do not contain an attachment.
